### PR TITLE
feat: add support for sec=krb5 mounting

### LINKED ISSUE #512: Kerberos ticket
Would it be possible to not go for username and password but to use a Kerberos ticket for the mount instead?

It would be the equivalent of a mount on the host as:
//server/share /mnt cifs multiuser,sec=krb5,user 0 0

Users will have access to it after a `kinit` to get a kerberos ticket. Equivalently the pod would need a kerberos ticket but that could be provided by an init script.

### PR DESCRIPTION
**Pull request by yerenkow**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Adding support for more secure mounting, without exposing password

**Which issue(s) this PR fixes**:
Fixes #512 

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
Adding support for passwordless mounting on linux using provided kerberos credential cache with tickets. 
```

### CODE REVIEW

**Comment by andyzhangx** [comment]
@yerenkow thanks for the contribution, could you sign the cla first?

File: pkg/smb/nodeserver.go [hunk 1]
```diff
@@ -422,3 +427,71 @@ func checkGidPresentInMountFlags(mountFlags []string) bool {
...
+		var krb5CcacheName = getKrb5CcacheName(credUid)
+		var krb5CcacheContent string
+		for k, v := range secrets {
+			switch strings.ToLower(k) {
+			case krb5CcacheName:
+				krb5CcacheContent = v
+			}
+		}
+		if krb5CcacheContent == "" {
+			return false, status.Error(codes.Internal, fmt.Sprintf("Empty kerberos cache in key %s", krb5CcacheName))
```
Comment (andyzhangx) [inline pkg/smb/nodeserver.go:479]: `codes.InvalidArgument`

File: pkg/smb/nodeserver.go [hunk 2]
```diff
@@ -422,3 +427,71 @@ func checkGidPresentInMountFlags(mountFlags []string) bool {
...
+		}
+		content, err := base64.StdEncoding.DecodeString(krb5CcacheContent)
+		if err != nil {
+			return false, status.Error(codes.Internal, fmt.Sprintf("Malformed kerberos cache in key %s, expected to be in base64 form: %v", krb5CcacheName, err))
```
Comment (andyzhangx) [inline pkg/smb/nodeserver.go:483]: codes.InvalidArgument

File: docs/driver-parameters.md [hunk 3]
```diff
@@ -34,6 +34,41 @@ nodeStageSecretRef.namespace | namespace where the secret is | k8s namespace  |
...
+export KRB5CCNAME=/tmp/ccache # Use temporary file for the cache 
+kinit USERNAME # Log in into domain
+kvno cifs/SERVER_NAME # Acquire ticket for the needed share, it'll be written to the cache file
+CCACHE=$(base64 -w 0 $KRB5CCNAME) # Get Base64-encoded cache
+```
+
+And passing the actual ticket to the secret, instead of the password.
+Note that key for the ticket has included credential id, that must match exactly `cruid=` mount flag.
+```console
+kubectl create secret generic smbcreds-krb5 --from-literal username=USERNAME --from-literal krb5cc_1000=$CCACHE
```
Comment (andyzhangx) [inline docs/driver-parameters.md:66]: from the code, I find only `krb5cc_1000` is used, `username` is not used.

File: pkg/smb/nodeserver.go [hunk 4]
```diff
@@ -422,3 +427,71 @@ func checkGidPresentInMountFlags(mountFlags []string) bool {
...
+
+func getKrb5CacheFileName(credUid int) string {
+	return fmt.Sprintf("%s%s%d", krb5CacheDirectory, krb5Prefix, credUid)
+}
+
+func ensureKerberosCache(mountFlags []string, secrets map[string]string) (bool, error) {
+	var securityIsKerberos = checkSecurityIsKerberos(mountFlags)
+	if securityIsKerberos {
+		if _, err := os.Stat(krb5CacheDirectory); !os.IsNotExist(err) {
+			// path/to/whatever exists
```
Comment (andyzhangx) [inline pkg/smb/nodeserver.go:462]: if it's other error, then it should return error directly.

File: pkg/smb/nodeserver.go (hunk 1 above)
Comment (yerenkow) [reply pkg/smb/nodeserver.go:479]: Done

File: pkg/smb/nodeserver.go (hunk 4 above)
Comment (yerenkow) [reply pkg/smb/nodeserver.go:462]: Done

**Comment by yerenkow** [comment]
@andyzhangx Thanks for review - I've updated and signed cla.

**Review by andyzhangx — COMMENTED** [review]
/ok-to-test

File: pkg/smb/nodeserver.go [hunk 5]
```diff
@@ -422,3 +428,88 @@ func checkGidPresentInMountFlags(mountFlags []string) bool {
 	}
 	return false
 }
+
+func checkSecurityIsKerberos(mountFlags []string) bool {
```
Comment (andyzhangx) [inline pkg/smb/nodeserver.go:432]: rename as `hasKerberosMountOption`

File: docs/driver-parameters.md [hunk 6]
```diff
@@ -34,6 +34,42 @@ nodeStageSecretRef.namespace | namespace where the secret is | k8s namespace  |
 kubectl create secret generic smbcreds --from-literal username=USERNAME --from-literal password="PASSWORD"
 ```
 
+### Kerberos ticket support for Linux
+#### All the pre-requisites to make it work
+ - Each node must have kerberos support set up, and cifs-utils installed
+ - `/var/lib/kubelet/kerberos/` directory must exist - it'll contain kerberos credential cache files for different users.
+   - This is directory that is shared by smb container and the host.
+ - Each node will be responsible for cleaning that directory as admin see fit. **Unmount does not delete the cache file.**
```
Comment (andyzhangx) [inline docs/driver-parameters.md:42]: can you rephrase these words in chatGpt, this sounds like non native english

File: deploy/example/pv-smb-krb5.yaml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    - noserverino  # required to prevent data corruption
+  csi:
+    driver: smb.csi.k8s.io
+    readOnly: false
+    # volumeHandle format: {smb-server-address}#{sub-dir-name}#{share-name}
+    # make sure this value is unique for every share in the cluster
+    volumeHandle: smb-server.default.svc.cluster.local/share##
+    volumeAttributes:
+      source: "//smb-server.default.svc.cluster.local/share"
+    nodeStageSecretRef:
```
Comment (andyzhangx) [inline deploy/example/pv-smb-krb5.yaml:38]: I think using a storage class example is better

File: docs/driver-parameters.md (hunk 6 above)
Comment (yerenkow) [reply docs/driver-parameters.md]: Done

File: pkg/smb/nodeserver.go (hunk 5 above)
Comment (yerenkow) [reply pkg/smb/nodeserver.go]: Done

File: deploy/example/pv-smb-krb5.yaml (hunk 7 above)
Comment (yerenkow) [reply deploy/example/pv-smb-krb5.yaml]: Done

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Comment by andyzhangx** [comment]
@yerenkow when the kerberos ticket is expired, how does this driver handle this issue? do we need to remount the pv?
